rules/utils/normalize-range.js

```diff
@@ -0,0 +1,10 @@
+module.exports = node => {
+	let {range: [start, end] = [node.start, node.end], type, tail} = node;
+
+	if (type === 'TemplateElement') {
+		start += 1;
+		end -= tail ? 1 : 2;
+	}
+
+	return [start, end];
+};
```

Comment:
I'm not sure we need this, this suppose to only work for case in `no-hex-escape` rule.

The range of `TemplateElement` is not wrong, we just don't want replace `` ` ``, `${` and `}`.

For example we have a `Literal` string, if we want replace string inside quotes, we may `replaceTextRange([start+1, end - 1], 'replacement')`, but if we want replace the whole string, we may do this `replaceText(range, QUOTED_STRING)`, [we do this a lot](https://github.com/sindresorhus/eslint-plugin-unicorn/search?q=quoteString&unscoped_q=quoteString).

Reply:
Suggest move these to `no-hex-escape.js`, in other rules we just use `node.range`.

What do you think?

Reply:
Or should I add a second param for this util?

Reply:
> in other rules we just use node.range

`range` is not always available, that's why I add a default value for it: `[node.start, node.end].`

Reply:
I don't think we need this. The string case in `no-hex-escape.js`, we are using `node.raw` https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/no-hex-escape.js#L21 . 

So quotes are kept, but if we use `node.value`, we can't say we need `normalize` the range, we just fix the range ourself( or add quotes like we already do in other rules).

Reply:
Make sense?

Reply:
Then should `normalize-range` be kept? As @sindresorhus 's [suggestion](#issuecomment-584121177), we should not use `node.start/end` directly, but `node.range` is not available at https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/escape-case.js#L55

Reply:
It would be:

```js
module.exports = node => {
	const {range: [start, end] = [node.start, node.end], type, tail} = node;
	return [start, end];
};
```

Reply:
> > in other rules we just use node.range
> 
> `range` is not always available, that's why I add a default value for it: `[node.start, node.end].`

Sorry, I didn't see this until now, can you explain more, or give me a example?

Reply:
> It would be:
> 
> ```js
> module.exports = node => {
> 	const {range: [start, end] = [node.start, node.end]} = node;
> 	return [start, end];
> };
> ```

I'm going to take a look at this, thanks.

Reply:
I had a quick look ,  this seems a different AST from `regexpp`

https://github.com/mysticatea/regexpp/blob/d1808145b2861738747bb2e742537d87527a6871/src/ast.ts#L60-L71

there is no range, so we should use `.start` and `.end` here, maybe add a comment there. AST in other cases all has `range`, right?

Reply:
I suggest remove `normalize-range`, add `TemplateElement` logic to `no-hex-escape.js`, keep using `.start` and `.end` in AST from `regexpp`, and use `.range` directly in  other place.

Reply:
Do you mean keep `normalize-range.js` with:

```js
module.exports = node => {
    // there is no `range` property in AST from `regexpp`
    // reference: https://github.com/mysticatea/regexpp/blob/master/src/ast.ts#L60-L71
	const {range: [start, end] = [node.start, node.end]} = node;
	return [start, end];
};
```

Reply:
I'm aware we would meet other edge cases in the future, keeping `normalize-range` could help.

Reply:
I mean remove `normalize-range.js`, 

put this comment in 

```js
// there is no `range` property in AST from `regexpp`
// reference: https://github.com/mysticatea/regexpp/blob/master/src/ast.ts#L60-L71
```
in https://github.com/JounQin/eslint-plugin-unicorn/blob/patch-1/rules/escape-case.js#L56 , and use `.start`/`.end` there.

Other place use `node.range` directly

Reply:
> I'm afraid that we would meet other edge cases in the future, keeping `normalize-range` could help.

But if we keep this `normalize-range`, we have to use it whenever we need access `.range`, if there is other case we meet later, we can handle later.

Reply:
@fisker Fine enough.